### DISCUSSION TITLE
Is it possible to develop a new backend server like Golang?

### DISCUSSION THREAD

**Discussion by liweijian** [coder/code-server#6130] [Q&A] [answered]
It sounds a lot like we are using TypeScript+Node as our backend server:
https://github.com/coder/code-server/blob/main/src/node/main.ts

I was wondering if it is possible to develop a new backend server using another language, like Golang? Is the protocol easy to port to Golang?

**Comment by code-asher** [comment, marked as answer]
It should be possible to use a different language.  We pass the 
web socket to VS Code via Node's IPC so we would need some way to 
do the same in Golang.  No need to port any protocol; we just need 
to hook up the web socket to the VS Code process (well and pass
through any other HTTP requests as well).

I am not sure how much value there would be in doing this though; 
code-server is a pretty thin wrapper around VS Code which itself 
of course uses Node so most everything would still be Node anyway.

**Reply by liweijian** [reply]
@code-asher Awesome thanks for your help. Out of curiosity, do you happen to know where could I find documents about the IPC between Node and VS Code?

**Reply by code-asher** [reply]
It uses Node's `child_process.send()` which is documented here: https://nodejs.org/api/child_process.html#subprocesssendmessage-sendhandle-options-callback

There are a few Golang packages for IPC with Node so maybe one of those could help.

There might be a better way to do this though.  In VS Code there is a web server.  The way code-server works is it imports the endpoint handlers directly and calls them from its own web server which is possible since both are Node.  

Initially I was thinking the VS Code portions of the web server would be re-implemented in Golang and then for the main VS Code process you could spawn it without the web server and pass through the web socket via IPC (I think this will require some patching on the VS Code side as well).

But you could also spawn VS Code including its web server (maybe spawn it on a Unix socket but a port would work too) and then proxy everything between the Golang and VS Code web servers.  That way you would not have to deal with passing around a web socket via IPC or re-implementing anything.  The downside is technically needing to run two web servers (one in Golang, one in Node) and proxying between them although I doubt there is much overhead.

**Reply by code-asher** [reply]
If you do take the IPC route take a look at these: 

https://github.com/microsoft/vscode/blob/1.77.1/src/vs/server/node/extensionHostConnection.ts#L259-L262
https://github.com/microsoft/vscode/blob/b7886d7461186a5eac768481578c1d7ca80e2d21/src/vs/server/node/extensionHostConnection.ts#L44

The way to spawn VS Code processes is to spawn `bootstrap-fork.js` with some flags and the `VSCODE_AMD_ENTRYPOINT` environment variable.
